File: analysis/analyze_audio.py
import sys
import json
import speech_recognition as sr
from pydub import AudioSegment
from textblob import TextBlob
import textwrap
import re
import os

def format_text(text, width=80):
    sentences = text.replace('. ', '.\n').replace('? ', '?\n').replace('! ', '!\n')
    return textwrap.fill(sentences, width=width)

def lexical_diversity(text):
    words = re.findall(r'\b\w+\b', text.lower())
    total_words = len(words)
    unique_words = len(set(words))
    score = unique_words / total_words if total_words else 0
    return score, total_words, unique_words

def sentence_complexity(text):
    sentences = re.split(r'[.!?]', text)
    sentences = [s.strip() for s in sentences if s.strip()]
    num_sentences = len(sentences)
    words = re.findall(r'\b\w+\b', text)

    avg_sentence_length = len(words) / num_sentences if num_sentences else 0

    conjunctions = ['and', 'but', 'or', 'because', 'although', 'since', 'while', 'if', 'when', 'though']
    conjunction_count = sum(text.lower().count(c) for c in conjunctions)

    feedback = "✅ Balanced complexity." if avg_sentence_length > 12 and conjunction_count >= num_sentences else "⚠️ Simple or flat sentence structure."

    return avg_sentence_length, conjunction_count, feedback

def output_json(data):
    """Helper function to output JSON and handle errors"""
    try:
        json_str = json.dumps(data)
        print(json_str)
        sys.stdout.flush()  # Ensure the output is flushed immediately
    except Exception as e:
        error_data = {
            "error": "Failed to serialize output",
            "details": str(e)
        }
        print(json.dumps(error_data))
        sys.stdout.flush()
        sys.exit(1)

def main():
    if len(sys.argv) < 2:
        output_json({"error": "No audio file path provided"})
        sys.exit(1)

    audio_file = sys.argv[1]
    language = 'en-US'
    recognizer = sr.Recognizer()

    try:
        # Preprocess audio
        audio = AudioSegment.from_file(audio_file)
        audio = audio.set_frame_rate(16000)
        temp_wav_file = "temp.wav"
        audio.export(temp_wav_file, format="wav")

        with sr.AudioFile(temp_wav_file) as source:
            audio_data = recognizer.record(source)

        # Transcription
        text = recognizer.recognize_google(audio_data, language=language)
        formatted_text = format_text(text)

        # Sentiment
        blob = TextBlob(text)
        polarity = blob.sentiment.polarity
        subjectivity = blob.sentiment.subjectivity

        # Lexical Diversity
        diversity, total_words, unique_words = lexical_diversity(text)
        lex_pattern = "✅ Diverse vocabulary." if diversity > 0.5 else "⚠️ Repeated words or less variety."

        # Sentence Complexity
        avg_len, conj_count, complexity_feedback = sentence_complexity(text)

        # Clean up temp file
        if os.path.exists(temp_wav_file):
            os.remove(temp_wav_file)

        # Output as JSON with consistent property names
        result = {
            "transcribedText": formatted_text,
            "polarityScore": polarity,
            "subjectivityScore": subjectivity,
            "totalWords": total_words,
            "uniqueWords": unique_words,
            "diversityScore": diversity,
            "diversityFeedback": lex_pattern,
            "avgSentenceLength": avg_len,
            "conjunctionCount": conj_count,
            "complexityFeedback": complexity_feedback
        }
        output_json(result)
    except sr.UnknownValueError:
        output_json({"error": "Speech Recognition could not understand audio"})
        sys.exit(1)
    except sr.RequestError as e:
        output_json({"error": f"Could not request results from Speech Recognition service: {e}"})
        sys.exit(1)
    except Exception as e:
        output_json({"error": f"An error occurred: {e}"})
        sys.exit(1)

if __name__ == "__main__":
    main()

File: analysis/analyze_audio.js
const fs = require('fs');
const path = require('path');
const { spawn } = require('child_process');

// Text analysis functions
function lexicalDiversity(text) {
    const words = text.toLowerCase().match(/\b\w+\b/g) || [];
    const totalWords = words.length;
    const uniqueWords = new Set(words).size;
    const score = totalWords > 0 ? uniqueWords / totalWords : 0;
    return { score, totalWords, uniqueWords };
}

function sentenceComplexity(text) {
    const sentences = text.split(/[.!?]/).filter(s => s.trim());
    const numSentences = sentences.length;
    const words = text.toLowerCase().match(/\b\w+\b/g) || [];
    
    const avgSentenceLength = numSentences > 0 ? words.length / numSentences : 0;
    
    const conjunctions = ['and', 'but', 'or', 'because', 'although', 'since', 'while', 'if', 'when', 'though'];
    const conjunctionCount = conjunctions.reduce((count, conj) => 
        count + (text.toLowerCase().match(new RegExp(`\\b${conj}\\b`, 'g')) || []).length, 0);
    
    const feedback = avgSentenceLength > 12 && conjunctionCount >= numSentences 
        ? "✅ Balanced complexity." 
        : "⚠️ Simple or flat sentence structure.";
    
    return { avgSentenceLength, conjunctionCount, feedback };
}

function formatText(text, width = 80) {
    return text.replace(/\. /g, '.\n').replace(/\? /g, '?\n').replace(/! /g, '!\n');
}

// Simple sentiment analysis (basic implementation)
function analyzeSentiment(text) {
    const positiveWords = ['good', 'great', 'excellent', 'amazing', 'wonderful', 'happy', 'love', 'like'];
    const negativeWords = ['bad', 'terrible', 'awful', 'hate', 'dislike', 'sad', 'angry', 'frustrated'];
    
    const words = text.toLowerCase().match(/\b\w+\b/g) || [];
    const positiveCount = words.filter(word => positiveWords.includes(word)).length;
    const negativeCount = words.filter(word => negativeWords.includes(word)).length;
    
    const polarity = (positiveCount - negativeCount) / Math.max(words.length, 1);
    const subjectivity = Math.min((positiveCount + negativeCount) / Math.max(words.length, 1), 1);
    
    return { polarity, subjectivity };
}

function main() {
    const audioFile = process.argv[2];
    
    if (!audioFile) {
        console.error(JSON.stringify({ error: "No audio file path provided" }));
        process.exit(1);
    }
    
    // For now, return a mock result since we need Google Speech API integration
    // In production, you'd integrate with Google Cloud Speech-to-Text API
    const mockText = "This is a sample transcription of the audio file. The analysis would be performed on the actual transcribed text.";
    
    try {
        const formattedText = formatText(mockText);
        const { polarity, subjectivity } = analyzeSentiment(mockText);
        const { score: diversity, totalWords, uniqueWords } = lexicalDiversity(mockText);
        const { avgSentenceLength, conjunctionCount, feedback: complexityFeedback } = sentenceComplexity(mockText);
        
        const lexPattern = diversity > 0.5 ? "✅ Diverse vocabulary." : "⚠️ Repeated words or less variety.";
        
        const result = {
            transcribedText: formattedText,
            polarityScore: polarity,
            subjectivityScore: subjectivity,
            totalWords,
            uniqueWords,
            diversityScore: diversity,
            diversityFeedback: lexPattern,
            avgSentenceLength,
            conjunctionCount,
            complexityFeedback
        };
        
        console.log(JSON.stringify(result));
    } catch (error) {
        console.error(JSON.stringify({ error: `An error occurred: ${error.message}` }));
        process.exit(1);
    }
}

if (require.main === module) {
    main();
}

module.exports = { lexicalDiversity, sentenceComplexity, analyzeSentiment, formatText }; 